src/mds/Server.cc

```diff
@@ -6290,6 +6404,156 @@ void Server::handle_client_removexattr(MDRequestRef& mdr)
+	      // we've found a new layout at this inode
+	      return {curr_layout, layout_xattr_info_t::InheritanceStatus::SET};
+          } else {
+	      return {curr_layout, layout_xattr_info_t::InheritanceStatus::INHERITED};
+          }
+        }
+
+        if (cur->is_root()) {
+          break;
+	}
+
+        cur = cur->get_projected_parent_dir()->get_inode();
+      }
+      mds->clog->error() << "no layout found at root dir!";
+      ceph_abort("no layout found at root dir! something is really messed up with layouts!");
+    };
+
+    const auto lxi = get_inherited_layout(cur);
+
+    if (name == "ceph.dir.layout"sv) {
+      *css << "{\"stripe_unit\": " << lxi.layout.stripe_unit
+	   << ", \"stripe_count\": " << lxi.layout.stripe_count
+	   << ", \"object_size\": " << lxi.layout.object_size
+	   << ", \"pool_name\": ";
+      mds->objecter->with_osdmap([lxi, &css](const OSDMap& o) {
+	  *css << "\"";
+          if (o.have_pg_pool(lxi.layout.pool_id)) {
+	    *css << o.get_pool_name(lxi.layout.pool_id);
+	  }
+	  *css << "\"";
+	});
+      *css << ", \"pool_id\": " << (uint64_t)lxi.layout.pool_id;
+      if (lxi.layout.pool_ns.length()) {
+	*css << ", \"pool_namespace\": \"" << lxi.layout.pool_ns << "\"";
+      }
+
+      *css << ", \"inheritance\": \"@"
+	   << layout_xattr_info_t::status_to_string(lxi.status) << "\"}";
+    } else if (name == "ceph.dir.layout.stripe_unit"sv) {
+      *css << lxi.layout.stripe_unit;
+    } else if (name == "ceph.dir.layout.stripe_count"sv) {
+      *css << lxi.layout.stripe_count;
+    } else if (name == "ceph.dir.layout.object_size"sv) {
+      *css << lxi.layout.object_size;
+    } else if (name == "ceph.dir.layout.pool_name"sv) {
+      mds->objecter->with_osdmap([lxi, &css](const OSDMap& o) {
+          if (o.have_pg_pool(lxi.layout.pool_id)) {
+	    *css << o.get_pool_name(lxi.layout.pool_id);
+	  }
+        });
+    } else if (name == "ceph.dir.layout.pool_id"sv) {
+      *css << (uint64_t)lxi.layout.pool_id;
+    } else if (name == "ceph.dir.layout.pool_namespace"sv) {
+      *css << lxi.layout.pool_ns;
+    } else {
+      r = -CEPHFS_ENODATA; // no such attribute
+    }
+  } else if (name.substr(0, 12) == "ceph.dir.pin"sv) {
+    if (name == "ceph.dir.pin"sv) {
+      *css << cur->get_projected_inode()->export_pin;
+    } else if (name == "ceph.dir.pin.random"sv) {
+      *css << cur->get_projected_inode()->export_ephemeral_random_pin;
+    } else if (name == "ceph.dir.pin.distributed"sv) {
+      *css << cur->get_projected_inode()->export_ephemeral_distributed_pin;
+    } else {
+      // otherwise respond as invalid request
+      // since we only handle ceph vxattrs here
+      r = -CEPHFS_ENODATA; // no such attribute
+    }
+  } else {
+    // otherwise respond as invalid request
+    // since we only handle ceph vxattrs here
+    r = -CEPHFS_ENODATA; // no such attribute
+  }
+
+  if (r == 0) {
+    std::string s{css->strv()};
+    ENCODE_START(1, 1, bl);
+    encode(s, bl);
```

Comment:
```suggestion
    ENCODE_START(1, 1, bl);
    encode(css->strv(), bl);
```

Should work? To avoid unnecessary string allocation.

Reply:
ack; unnecessary string allocation omitted